### ISSUE TITLE
Property mapping fails when one property name matches the starting string portion of another longer property name

### ISSUE THREAD

**Issue by elgonzo** [dotnet/command-line-api#1195]
With the current nuget package 2.0.0-beta1.20574.7, the following code example runs fine:
```cs
public class MySettings
{
	public IEnumerable<string> Exclusions { get; set; }
	public string ExclusionsFile { get; set; }
}


public class Test
{
    public static void RunTest()
    {
        var rootCmd = new RootCommand()
        {
            new Option<IEnumerable<string>>("--Exclusions"),
            new Option<string>("--ExclusionsFile")
        };

        rootCmd.Handler = CommandHandler.Create(
            (MySettings s) =>
            {
                Console.WriteLine($"Exclusions = {s.Exclusions}");
                Console.WriteLine($"ExclusionsFile = {s.ExclusionsFile}");
            }
        );

        var args = new[] { "--Exclusions", "1" };

        rootCmd.Invoke(args);
    }
}
```

However, the same code fails using the current master branch (last commit at the time of me cloning/checking out the master branch: https://github.com/dotnet/command-line-api/commit/fd3cf0cdd7c39c745ad0fff8df542726301d3177), throwing an exception:
```
Unhandled exception: System.ArgumentException: Value System.Collections.Generic.List`1[System.String] (System.Collections.Generic.List`1[System.String]) must be an instance of type System.String
   at System.CommandLine.Binding.BoundValue..ctor(Object value, IValueDescriptor valueDescriptor, IValueSource valueSource) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\BoundValue.cs:line 16
   at System.CommandLine.Binding.BindingContext.TryBindToScalarValue(IValueDescriptor valueDescriptor, IValueSource valueSource, BoundValue& boundValue) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\BindingContext.cs:line 116
   at System.CommandLine.Binding.ModelBinder.GetBoundValue(IValueSource valueSource, BindingContext bindingContext, IValueDescriptor valueDescriptor, Boolean includeMissingValues, Type parentType) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\ModelBinder.cs:line 248
   at System.CommandLine.Binding.ModelBinder.GetBoundValues(IDictionary`2 bindingSources, BindingContext bindingContext, IReadOnlyList`1 valueDescriptors, Boolean enforceExplicitBinding, Type parentType, Boolean includeMissingValues) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\ModelBinder.cs:line 200
   at System.CommandLine.Binding.ModelBinder.UpdateInstanceInternalNotifyIfNonDefaultsUsed[T](T instance, BindingContext bindingContext) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\ModelBinder.cs:line 136
   at System.CommandLine.Binding.ModelBinder.InstanceFromSpecificConstructor(BindingContext bindingContext, ConstructorDescriptor constructor, IReadOnlyList`1 boundValues, Boolean& nonDefaultsUsed) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\ModelBinder.cs:line 126
   at System.CommandLine.Binding.ModelBinder.CreateInstanceInternal(BindingContext bindingContext) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\ModelBinder.cs:line 80
   at System.CommandLine.Binding.ModelBinder.GetBoundValue(IValueSource valueSource, BindingContext bindingContext, IValueDescriptor valueDescriptor, Boolean includeMissingValues, Type parentType) in X:\BallPit\System.CommandLine\src\System.CommandLine\Binding\ModelBinder.cs:line 264
   ...
```

While the argument value "1" is correctly parsed and converted into a collection with the single string "1", the current master branch attempts to assign this collection to the property `ExclusionsFile` instead to `Exclusions`. Fortunately, in this case `ExclusionsFile` is of an incompatible type (`string` instead of `IEnumerable<string>`), and setting the wrong property doesn't succeed. However, if `ExclusionsFile` were of the same type as `Exclusions`, setting the wrong property would have succeeded without throwing an exception.

The problem seems to be with the property names, and not so much with the option names themselves. Because, change the property name `ExclusionsFile` to `XExclusionsFile`, and the problem goes away. However, keep the `ExclusionsFile` property name and change only the option name from `--ExclusionsFile` to `--SomethingElse`, and the problem still occurs.

**Comment by elgonzo** [comment]
The bug seems to have been introduced with merged pull request https://github.com/dotnet/command-line-api/pull/1080 which drastically changed the behavior Binder.IsMatch method (https://github.com/dotnet/command-line-api/commit/269a1acffc2b1bbf03f65100df54d8a3bd32e942#diff-94745ece8d7a189347723c99bdf5b7856e30d3fb5ca505d7d4b88c2c0219978f).

**Comment by paulomorgado** [comment]
@elgonzo,

Would you be so kind to submit additional unit test cases for this?

/cc @jonsequitur

**Comment by elgonzo** [comment]
> Would you be so kind to submit additional unit test cases for this?

Just to make clear, the current behavior of `Binder.IsMatch` is not intentional, correct?
If that's the case, i can cook up some unit test cases for this.

**Comment by paulomorgado** [comment]
I don't know if the previous behavior was intentional. 😄 

Let's see what @jonsequitur has to say about it.

**Comment by elgonzo** [comment]
Me neither. :)

If you look at my pull request, (https://github.com/dotnet/command-line-api/pull/1200), you'll notice that one unit test fails: `Method_parameters_on_the_invoked_method_are_bound_to_matching_option`. This unit test was only added in early January. That makes it kind of looking intentional, i guess, in which case the solution to the issue i reported is not done with the simple modification proposed in my pull request, i am afraid :(

**Comment by elgonzo** [comment]
Also worth noting is the fact that the latest version as released and obtainable via nuget (2.0.0-beta1.20574.7) still features the "old" behavior of `Binder.IsMatch`. So, the change of `Binder.IsMatch` to the current behavior in the main branch is a breaking change that has the potential to break code consuming this package when the developers update System.CommandLine when the next version is being published via nuget...
